extensions/amp-consent/0.1/consent-state-manager.js

```diff
@@ -198,80 +204,86 @@ export class ConsentInstance {
   /**
    * Update the local consent state list
    * @param {!CONSENT_ITEM_STATE} state
+   * @param {string=} consentStr
    */
-  update(state) {
-    if (!isEnumValue(CONSENT_ITEM_STATE, state)) {
-      state = CONSENT_ITEM_STATE.UNKNOWN;
-    }
-
-    if (state == CONSENT_ITEM_STATE.DISMISSED) {
-      this.localValue_ = this.localValue_ || CONSENT_ITEM_STATE.UNKNOWN;
-      return;
+  update(state, consentStr) {
+    const localStateValue =
+        this.localConsentInfo_ && this.localConsentInfo_['consentState'];
+    const localConsentStr =
+        this.localConsentInfo_ && this.localConsentInfo_['consentString'];
+    const normalizedState = normalizeConsentStateValue(state, localStateValue);
+    if (consentStr === undefined && localConsentStr) {
+      consentStr = localConsentStr;
     }
+    const newConsentInfo = constructConsentInfo(normalizedState, consentStr);
+    const oldConsentInfo = this.localConsentInfo_;
+    this.localConsentInfo_ = newConsentInfo;
 
-    if (state == CONSENT_ITEM_STATE.NOT_REQUIRED) {
-      if (!this.localValue_ || this.localValue_ == CONSENT_ITEM_STATE.UNKNOWN) {
-        this.localValue_ = CONSENT_ITEM_STATE.NOT_REQUIRED;
-      }
-      return;
-    }
-
-    if (state === this.localValue_) {
-      return;
-    }
-
-    this.localValue_ = state;
-
-    if (state == CONSENT_ITEM_STATE.UNKNOWN) {
-      return;
+    if (!consentInfoEquals(newConsentInfo, oldConsentInfo)) {
+      this.updateStoredValue_(newConsentInfo);
+      // TODO(@zhouyx): Need force update to update timestamp
     }
+  }
 
-    const value = (state == CONSENT_ITEM_STATE.ACCEPTED);
+  /**
+   * Write the new value to localStorage and send updateHrefRequest
+   * @param {!ConsentInfoDef} consentInfo
+   */
+  updateStoredValue_(consentInfo) {
     this.storagePromise_.then(storage => {
-      if (state != this.localValue_) {
-        // If state has changed. do not store.
+      if (!consentInfoEquals(consentInfo, this.localConsentInfo_)) {
```

Comment:
Exactly. It's very unlikely this will happen, but it's possible that user update their consent again before the promise resolve. There's a unit test for this case https://github.com/ampproject/amphtml/blob/master/extensions/amp-consent/0.1/test/test-consent-state-manager.js#L177